src/native/libs/System.Security.Cryptography.Native.Apple/pal_swiftbindings.swift

```diff
@@ -350,3 +358,202 @@ public func AppleCryptoNative_HKDFDeriveKey(
+        return -1
+    }
+
+    let box = Unmanaged<HashBox>.fromOpaque(ctx).takeUnretainedValue()
+    let source = Data(bytesNoCopy: pBuf, count: Int(cBuf), deallocator: Data.Deallocator.none)
+    var hash = box.value
+    hash.update(data: source)
+    box.value = hash
+    return 1
+}
+
+@_silgen_name("AppleCryptoNative_DigestReset")
+@available(iOS 13, tvOS 13, *)
+public func AppleCryptoNative_DigestReset(ctx: UnsafeMutableRawPointer?) -> Int32 {
+    guard let ctx else {
+        return -1
+    }
+
+    let box = Unmanaged<HashBox>.fromOpaque(ctx).takeUnretainedValue()
+
+    switch box.value {
+        case is Insecure.MD5:
+            box.value = Insecure.MD5()
+            return 1
+        case is Insecure.SHA1:
+            box.value = Insecure.SHA1()
+            return 1
+        case is SHA256:
+            box.value = SHA256()
+            return 1
+        case is SHA384:
+            box.value = SHA384()
+            return 1
+        case is SHA512:
+            box.value = SHA512()
+            return 1
+        default:
+            return -2
+    }
+}
+
+@_silgen_name("AppleCryptoNative_DigestFinal")
+@available(iOS 13, tvOS 13, *)
+public func AppleCryptoNative_DigestFinal(ctx: UnsafeMutableRawPointer?, pOutput: UnsafeMutablePointer<UInt8>?, cbOutput: Int32) -> Int32 {
+    guard let ctx, let pOutput else {
+        return -1
+    }
+
+    let box = Unmanaged<HashBox>.fromOpaque(ctx).takeUnretainedValue()
+    let destination = UnsafeMutableRawBufferPointer(start: pOutput, count: Int(cbOutput))
+
+    let hash = box.value.finalize()
+    let copied = hash.withUnsafeBytes { digest in
+        return digest.copyBytes(to: destination) == digest.count
+    }
+
+    if (!copied) {
+        return -1
+    }
+
+    return AppleCryptoNative_DigestReset(ctx: ctx)
+}
+
+@_silgen_name("AppleCryptoNative_DigestFree")
+@available(iOS 13, tvOS 13, *)
+public func AppleCryptoNative_DigestFree(ptr: UnsafeMutableRawPointer?) {
+    if let ptr {
+        Unmanaged<HashBox>.fromOpaque(ptr).release()
+    }
+}
+
+@_silgen_name("AppleCryptoNative_DigestClone")
+@available(iOS 13, tvOS 13, *)
+public func AppleCryptoNative_DigestClone(ctx: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
+    guard let ctx else {
+        return nil
+    }
+
+    let box = Unmanaged<HashBox>.fromOpaque(ctx).takeUnretainedValue()
+    let hash = box.value
+    let copy = hash
+    let cloneBox = HashBox(copy)
+    return Unmanaged.passRetained(cloneBox).toOpaque()
+}
+
+@_silgen_name("AppleCryptoNative_DigestCurrent")
+@available(iOS 13, tvOS 13, *)
+public func AppleCryptoNative_DigestCurrent(ctx: UnsafeMutableRawPointer?, pOutput: UnsafeMutablePointer<UInt8>?, cbOutput: Int32) -> Int32 {
+    guard let ctx, let pOutput else {
+        return -1
+    }
+
+    let box = Unmanaged<HashBox>.fromOpaque(ctx).takeUnretainedValue()
+    let destination = UnsafeMutableRawBufferPointer(start: pOutput, count: Int(cbOutput))
+    let hash = box.value.finalize()
```

Comment:
The docs for finalize say that the HashFunction should be discarded after the call to finalize, but this looks like it's treating it as if finalize is idempotent.

I'd expect

```swift
let hasher = box.value
let hash = hasher.finalize()
```

But maybe I don't understand either the value copy semantics or the way HashFunction actually works?

Reply:
"Value copy semantics"

Getting the value from the box is itself a copy. We would need to put the copy back in the box.

_So then why the explicit copy in Clone_?

We don't need it.